adapters/aduptech/aduptech.go

```diff
@@ -0,0 +1,163 @@
+	}
+
+	if extraInfo.TargetCurrency == "" {
+		return nil, fmt.Errorf("invalid extra info: TargetCurrency is empty, pls check")
+	}
+
+	bidder := &adapter{
+		endpoint:  config.Endpoint,
+		extraInfo: extraInfo,
+	}
+
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	for i := range request.Imp {
+		imp := &request.Imp[i]
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != a.extraInfo.TargetCurrency {
+
+			convertedValue, err := a.convertCurrency(imp.BidFloor, imp.BidFloorCur, reqInfo)
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			imp.BidFloorCur = a.extraInfo.TargetCurrency
+			imp.BidFloor = convertedValue
+		}
+	}
+
+	requestJSON, err := jsonutil.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method: http.MethodPost,
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+		ImpIDs: openrtb_ext.GetImpIDs(request.Imp),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+// convertCurrency attempts to convert a given value from the specified currency (cur) to the
+// target currency specified in the adapter's extraInfo. If the conversion directly to the
+// target currency fails due to a ConversionNotFoundError, it attempts an intermediate conversion
+// through USD. Returns the converted value or an error if conversion fails.
+func (a *adapter) convertCurrency(value float64, cur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	convertedValue, err := reqInfo.ConvertCurrency(value, cur, a.extraInfo.TargetCurrency)
+
+	if err != nil {
+		var convErr currency.ConversionNotFoundError
+		if !errors.As(err, &convErr) {
+			return 0, err
+		}
+
+		// try again by first converting to USD
+		// then convert to target_currency
+		convertedValue, err = reqInfo.ConvertCurrency(value, cur, "USD")
+
+		if err != nil {
+			return 0, err
+		}
+
+		convertedValue, err = reqInfo.ConvertCurrency(convertedValue, "USD", a.extraInfo.TargetCurrency)
+
+		if err != nil {
+			return 0, err
+		}
+	}
+	return convertedValue, nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		return nil, []error{&errortypes.BadInput{
+			Message: "Unexpected status code: 400. Run with request.debug = 1 for more info.",
+		}}
+	case http.StatusOK:
+		// we don't need to do anything here and this is just so we can use the default case for the unexpected status code
+		break
+	default:
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", responseData.StatusCode),
+		}}
+
+	}
```

Comment:
Please use the adapter utility functions `IsResponseStatusCodeNoContent` and `CheckResponseStatusCodeForErrors` found in `adapters/response.go` for processing the http status code.